### PR TITLE
Filter transform '!=' operation and 'preservegaps' option

### PR DESCRIPTION
**Pull request by etpinard**
peek 2017-04-12 15-11

### CODE REVIEW

**Comment by etpinard** [comment]
@bpostlethwaite this resolves two feature requests of yours from last December.

**Comment by bpostlethwaite** [comment]
nice! Will be interesting to integrate this back into Workspace

File: src/transforms/filter.js [hunk 1]
```diff
@@ -15,7 +15,7 @@ var axisIds = require('../plots/cartesian/axis_ids');
 var autoType = require('../plots/cartesian/axis_autotype');
 var setConvert = require('../plots/cartesian/set_convert');
 
-var INEQUALITY_OPS = ['=', '<', '>=', '>', '<='];
+var INEQUALITY_OPS = ['=', '!=', '<', '>=', '>', '<='];
```
Comment (alexcjohnson) [inline src/transforms/filter.js]: hmm, nonblocking as this isn't new, but seems funny to categorize `'='` in `INEQUALITY_OPS`. How about `COMPARISON_OPS`?

File: src/transforms/filter.js [hunk 2]
```diff
@@ -101,7 +102,17 @@ exports.attributes = {
...
             'the desired set elements.'
         ].join(' ')
-    }
+    },
+    preservegaps: {
+        valType: 'boolean',
+        dflt: false,
+        description: [
+            'Determines whether or not gaps in data arrays produced by the filter operation',
+            'are preserved or not.',
```
Comment (alexcjohnson) [inline src/transforms/filter.js]: whether or not... or not

File: src/transforms/filter.js [hunk 3]
```diff
@@ -101,7 +102,17 @@ exports.attributes = {
...
+            'Setting this to *true* might be useful when plotting a line chart',
+            'with `connectgaps` set to *true*.'
```
Comment (alexcjohnson) [inline src/transforms/filter.js]: wait, do you mean with `connectgaps` set to `false`? `connectgaps: true` would run connect right over the gaps you just preserved...

Comment (etpinard) [inline src/transforms/filter.js]: Oh right. Good eye. Thanks!

File: test/jasmine/tests/transform_multi_test.js [hunk 4]
```diff
@@ -66,23 +67,23 @@ describe('general transforms:', function() {
 
         traceOut = Plots.supplyTraceDefaults(traceIn, 0, layout);
 
-        expect(traceOut.transforms[0]).toEqual({
+        expect(traceOut.transforms[0]).toEqual(jasmine.objectContaining({
```
Comment (alexcjohnson) [inline test/jasmine/tests/transform_multi_test.js]: oh, didn't know about that one, cool!

File: test/jasmine/tests/transform_filter_test.js [hunk 5]
```diff
@@ -320,6 +321,22 @@ describe('filter transforms calc:', function() {
...
+                value: 0,
+                target: 'x'
+            }]
+        })]);
+
+        expect(out[0].x).toEqual([undefined, undefined, undefined, undefined, 1, 2, 3]);
+        expect(out[0].y).toEqual([undefined, undefined, undefined, undefined, 2, 3, 1]);
+        expect(out[0].marker.color).toEqual([undefined, undefined, undefined, undefined, 0.2, 0.3, 0.4]);
+    });
+
```
Comment (alexcjohnson) [inline test/jasmine/tests/transform_filter_test.js]: Are there any interesting cases to test with two filters applied in sequence having different `preservegaps` values?

Comment (etpinard) [inline test/jasmine/tests/transform_filter_test.js]: Not yet. I'll add one in a future commit :rocket:

File: test/jasmine/tests/transform_filter_test.js [hunk 6]
```diff
@@ -337,6 +337,104 @@ describe('filter transforms calc:', function() {
...
+        expect(out0[0].marker.color).toEqual([0.1, 0.2]);
+
+        var out1 = _transform([Lib.extendDeep({}, base, {
+            transforms: [transform1, transform0]
+        })]);
+
+        expect(out1[0].x).toEqual([undefined, undefined, undefined, 0, 1]);
+        expect(out1[0].y).toEqual([undefined, undefined, undefined, 1, 2]);
+        expect(out1[0].marker.color).toEqual([undefined, undefined, undefined, 0.1, 0.2]);
+    });
```
Comment (alexcjohnson) [inline test/jasmine/tests/transform_filter_test.js]: Awesome, thanks for those tests.

**Comment by alexcjohnson** [comment]
💃

**Pull request merged**